### PR TITLE
[JENKINS-70922] Remove Prototype `Ajax.Request` usage from `select.js`

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-70922](https://issues.jenkins.io/browse/JENKINS-70922).

### Testing done

Manually tested by clicking through the UI where it's used adding console.log to see what was using it.
Credentials were fine.
Ran `mvn test -Dtest=hudson.RelativePathTest` which provides test coverage (and it caught an issue with parameters originally)

Draft till ran through ATH

### Proposed changelog entries

- Entry 1: Issue, human-readable text
- […]

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7982"></a>

### CODE REVIEW

**Review by basil — COMMENTED** [review]
This was meant to be a starting point for further exploration, not a final solution. I merely sought to demonstrate that interleaved callbacks from refill operations were interfering with each other and causing a data race.

(1) First, I think we should determine what is firing the second change event. Does it make sense, or is it unnecessary? If we can eliminate that, the entire problem is solved.

(2) _If_ the second change event is necessary, then I think we should explore whether the refill process can be made thread-safe with traditional techniques (e.g., reducing the use of shared data).

Only if the second change event is necessary (1) and it cannot be made thread-safe (2) do I think we should implement mutual exclusion (3) as in my cheesy sleep and retry demonstration.

And if we are to implement mutual exclusion (3), my cheesy sleep and retry demonstration is certainly not a good way to do it. Doing a quick Google search, I found a handful of [mutex libraries](https://github.com/DirtyHairy/async-mutex). Or for a library-less option, anyone can implement a Lamport lock [at](https://medium.engineering/wait-dont-touch-that-a211832adc3a) [home](https://swizec.com/blog/a-fast-mutex-lamport-lock-with-javascript-promises/) (which is essentially a mathematically correct version of my cheesy sleep and retry demonstration).

**Comment by timja** [comment]
> (1) First, I think we should determine what is firing the second change event. Does it make sense, or is it unnecessary? If we can eliminate that, the entire problem is solved.

It's only a single event but there's multiple event handlers registered to the element

image

Verified by removing one of the event handlers, and it only runs once.

**Comment by timja** [comment]
By sticking on a breakpoint and looking at the callstack I tracked the second event listener to this code:
https://github.com/jenkinsci/credentials-plugin/blob/b96f366e7badeedaf69724991e44409be11a07b6/src/main/resources/lib/credentials/select/select.js#L274-L282

Hopefully it can just be safely removed.

**Comment by timja** [comment]
It can't be removed as-is, the steps from https://github.com/jenkinsci/credentials-plugin/pull/23 still reproduce the issue

**Comment by timja** [comment]
That works now, thoughts @basil?

(Will need ATH testing as well)

**Comment by basil** [comment]
This is still failing ATH.

**Comment by timja** [comment]
> This is still failing ATH.

Hmm it's a really weird one.

When running through ATH it selects the credential and then it gets reset to none super quickly.
If I pause the test and do it myself it's just fine =/

**Comment by timja** [comment]
Tests seem to mostly pass after: https://github.com/jenkinsci/jenkins/pull/7982/commits/9bee3b1b20e8b4f06df336f295470425e296c6d6

I'm having to use chrome locally as Firefox isn't working on newer versions which might be why some tests are failing.

image


Let's see how it goes on the pipeline

**Comment by basil** [comment]
> When running through ATH it selects the credential and then it gets reset to none super quickly.
> If I pause the test and do it myself it's just fine =/

That would be evidence that the data race still exists.

**Comment by basil** [comment]
> Yes that was resolved by [9bee3b1](https://github.com/jenkinsci/jenkins/commit/9bee3b1b20e8b4f06df336f295470425e296c6d6)

Aha, this was the real fix, and the change to `refillOnChange-listener-attached` in `war/src/main/webapp/scripts/hudson-behavior.js` does not seem to make any difference when testing. So please revert the changes to `hudson-behavior.js`. If that change makes sense for independent reasons (e.g., efficiency, correctness, etc), it may be proposed as a separate PR, but it does not seem related to the primary purpose of this PR.

**Review by basil — APPROVED** [review]
ATH is looking good on the latest revision. Nice job taking this across the finish line!

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!
